### ISSUE TITLE
Convolver auto-gain

### ISSUE THREAD

**Issue by future-figs** [wwmm/easyeffects#1309]
### EasyEffects Version

6.1.5

### What package are you using?

Fedora

### Distribution

Fedora 35

### Describe the bug

When using the convolver, an unpredictable amount of gain is applied to the IR kernel, and therefore the output signal. This makes it difficult to prevent digital clipping in EQ applications without manually reducing the gain to excess.

The convolver has auto-gain functionality built in. It first normalizes the IR kernel to just short of digital clipping, like a CD master [-1.0 +1.0]. Then, if the sum of squares is still less than 1.0, it increases the gain further. This increases the volume level of the kernel in most cases, and can introduce digital clipping even with both preamp sliders left at 0 dB.
https://github.com/wwmm/easyeffects/blob/d820fa80b3b929225a46fbee253dccdff55e2c78/src/convolver.cpp#L357-L389

It's easy to apply some negative gain with the input/output sliders, but there's no way to know how much is enough—cue obsessing over every little pop or crackle.

### Expected Behavior

I can think of two potential ways to improve this behavior:
1. Change the auto-gain to use an absolute-value sum instead (greatest of either channel). If this sum is <=1, then digital clipping is impossible, unless extra gain is applied on top etc. Ideal for EQ applications, but tends to reduce the volume level noticeably. I can submit a PR if this approach is favored.
2. Do nothing. The user can either scale their IR kernels appropriately ahead of time, or else play with the preamp sliders like they currently have to anyway.

**Comment by wwmm** [comment]
>This increases the volume level of the kernel in most cases, and can introduce digital clipping even with both preamp sliders left at 0 dB.

That is weird. None of the impulse responses I have tested lead to an autogain factor higher than 1. In one of my impulse files the output signal is amplified even if the kernel is scaled with an factor equal to `0.614267`. So something else is at play. Part of the amplification seems to be related to the convolution process.

**Comment by wwmm** [comment]
>Then, if the sum of squares is still less than 1.0, it increases the gain further

I am doing some tests and I do not see this happening. When the average power is less than one `std::min` forces the autogain factor to be equal to `1`.

**Comment by Mr-Figs** [comment]
> > Then, if the sum of squares is still less than 1.0, it increases the gain further
> 
> I am doing some tests and I do not see this happening. When the average power is less than one `std::min` forces the autogain factor to be equal to `1`.

You are right—I got it backwards in the original post. The auto-gain step will actually never increase the gain, only decrease or leave as-is (port normalization). So it's actually on the right track already! However, I still think that the sum-of-squares approach (I now realize it's really a root-mean-square) isn't fit for purpose here.

> > This increases the volume level of the kernel in most cases, and can introduce digital clipping even with both preamp sliders left at 0 dB.
> 
> That is weird. None of the impulse responses I have tested lead to an autogain factor higher than 1. In one of my impulse files the output signal is amplified even if the kernel is scaled with an factor equal to `0.614267`. So something else is at play. Part of the amplification seems to be related to the convolution process.

I think you may be forgetting the normalization step, which is not accounted for by the "autogain" value written to log. A logged autogain of 1.0 actually means that the IR kernel was amplified to the max.

**Comment by Mr-Figs** [comment]
It also might be worth putting together a simple pathological test-case. Consider the following IR kernel, same for both L/R channels: [0.5 0.5 0 0]

If we apply no gain, it will not produce digital clipping. Since all the samples are positive, we can test this by feeding it a DC stream of 1.0 samples. The result is (1.0 * 0.5) + (1.0 * 0.5) = 1.0; on the limit, but not clipping.

The `apply_kernel_autogain()` function will first normalize, by dividing each sample by the peak. This changes both channels to [1.0 1.0 0 0]. This kernel would produce clipping.

Next step is sum of squares across both channels. This is equal to 4.0, which is divided by 2 to get 2.0. Finally, the inverse of the square root of 2 is 0.7071, which is less than 1.0, so the autogain factor (as logged) is 0.7071. The resulting IR kernel is [0.7071 0.7071 0 0], which produces peak values of 1.4142 in the DC test described above, i.e. clipping. It is reduced in volume, but not enough to counteract the normalization step.

If we use the absolute-value-sum method instead, we get a different result. The sum for either channel is 2.0, so the autogain factor is 1 / 2.0 = 0.5. The IR kernel is then changed back to [0.5 0.5 0 0], which doesn't produce clipping.

**Comment by wwmm** [comment]
>If we apply no gain, it will not produce digital clipping. Since all the samples are positive, we can test this by feeding it a DC stream of 1.0 samples. The result is (1.0 * 0.5) + (1.0 * 0.5) = 1.0; on the limit, but not clipping.

I haven't calculated a convolution by hand in a while. So I am not seeing how this test case is reproducing a convolution calculation.

**Comment by Mr-Figs** [comment]
Something weird I hadn't noticed before—even with the auto-gain function totally bypassed, there is still some gain applied. Maybe this is what you were talking about before. The really strange thing is, it seems to depend on the signal. I'm testing with a "unity" IR kernel ( [1 0 0 0] for each channel) which shouldn't have any effect on the signal.

With the built-in white noise test signal, it gains about 1 dB (it jumps around a lot, so it's hard to tell exactly)
![image](https://user-images.githubusercontent.com/8206263/147981460-2ff5a011-d69b-4845-924b-d5fb1b7faa6c.png)

Meanwhile a 1kHz sine gains closer to 8 dB(!). Same for other frequencies:
![image](https://user-images.githubusercontent.com/8206263/147981630-0f226a76-466f-402e-936b-944d3d4ab610.png)

No point obsessing over the IR kernel's levels if the convolver is going to behave this way. I'll try to dig a little deeper but in the likely event I can't figure it out...then never mind, I guess :(

**Comment by wwmm** [comment]
I think that the convolution process itself is having some part in the amplification or attenuation of the signal amplitude. When we use an equalizer and we apply a big boost to the lower frequencies the overall signal level becomes higher. Consider now that a kernel is designed to create the same effect. I would expect the output signal to have higher amplitude because the convolution process was designed to give a huge boost to the lower frequencies.

Most of what I did in this convolver code was inspired by what I saw in another convolver implemented by someone with way more experience on this topic https://github.com/tomszilagyi/ir.lv2. Maybe there are more steps done there that I did not notice.

**Comment by Mr-Figs** [comment]
> > If we apply no gain, it will not produce digital clipping. Since all the samples are positive, we can test this by feeding it a DC stream of 1.0 samples. The result is (1.0 * 0.5) + (1.0 * 0.5) = 1.0; on the limit, but not clipping.
> 
> I haven't calculated a convolution by hand in a while. So I am not seeing how this test case is reproducing a convolution calculation.

Yeah, sorry, that was pretty vague. Let me try to explain better. Here's a diagram that shows one output sample (y[n]) being determined by multiple input samples (x) and the IR kernel (h):
![image](https://user-images.githubusercontent.com/8206263/147983072-a6dffd17-7130-4016-987f-40e71774db90.png)
Source: https://www.analog.com/media/en/technical-documentation/dsp-book/dsp_book_Ch6.pdf (page 117)

In order to maximize that one sample y[n] and induce clipping if possible, each sample in x should be (1) the same sign as the corresponding sample in h, and (2) at the max possible absolute value, i.e. ±1.0. This way, when we reach the summing node, all its inputs are as large and positive as possible, for the greatest possible sum. If that doesn't get the output to clip, then nothing will.

Since we treat ±1.0 as the maximum values of either signal, we're just multiplying the IR kernel samples by ±1.0, which is the same as just taking the absolute value of these samples. That's why the DC trick works for a non-negative IR kernel—no absolute-value step necessary.

I hope that helps clear up my earlier explanation.

**Comment by wwmm** [comment]
I see. In any case we can try other normalization methods. I have no reason to keep using the current one. Is the method you suggest the same the book uses for the windowed-sync low pass kernel https://www.dspguide.com/ch16/2.htm?

**Comment by Mr-Figs** [comment]
Seems similar, except without the absolute value step. So it works fine for an ideal lowpass filter with unity gain at DC, and equal or less at any other frequency. However, it would not work so well for a highpass filter (it would become a high-boost filter, right?).

Maybe you could even get the book's method to produce clipping if you fed the kernel backwards into the convolver. Just because any one pure tone is unity or attenuated, I'm not sure that clipping is prevented. I could try it out in Octave.

Jan 3, 2022 5:28:07 PM Wellington Wallace ***@***.***>:

**Comment by wwmm** [comment]
I have done some tests with a normalization based on the sum(with and without applying `std::fabs` on each sample value) and the output signal level gets ridiculously low. At least with the impulse files I use I have to put more than 20 dB of output amplification to bring the signal level close to the one in the convolver input. This is more annoying than dealing with things as they are now.

**Comment by bhack** [comment]
@wwmm I've found a way to reproduce the clipping.

E.g. If you load the wav from https://github.com/jaakkopasanen/AutoEq/tree/master/results/oratory1990/harman_over-ear_2018/HIFIMAN%20Edition%20XS

You can test e.g. Big Science track like this:
https://www.last.fm/music/Laurie+Anderson/Big+Science/Born,+Never+Asked

If you can try it on Spotify that has voulume/LUFS normalizzation it will not go to clip if you load the parametirc parametricEQ but with the convolver + autogain and the same negative pre-amp as the parametricEQ it is going to clip  > +1.0dBTP (you can test yourself connecting `x42-meter` and switch on `Compute True-Peak`).

Honestly the autoeq website is a little bit ambiguos in the Easyeffect convolver section:
https://github.com/jaakkopasanen/AutoEq#pulseeffects--easyeffects
> The gain required by parametric eq should be sufficient, maybe 0.5 dB of negative gain more. Depending on the version, you may need to rename the .wav file as .irs.

**Comment by wwmm** [comment]
>I've found a way to reproduce the clipping.

I can see it here. I am not sure if there is much that can be done about it. This autogain algorithm performance seems to depend a lot on the impulse being used. On some it is fine but on others the final level is too high.

For my particular use cases I usually have the autogain plugin and the limiter as the last plugins in the pipeline. So the soundcard is not going to clip. But without them the convolver with impulse autogain enabled can be problem.

**Comment by bhack** [comment]
It seems that we had also a parallel ticket in Autoeq https://github.com/jaakkopasanen/AutoEq/issues/388

**Comment by bhack** [comment]
Do you think that we could add at least a dbtp peak in the UI near the meters (with a reset button)?

At least it will help users to manually adjust the negative input gain sliders on the fly on real tracks/session without needing to use external tools (e.g. x42 meters) just for the task to adapt the input gain.

**Comment by wwmm** [comment]
>Do you think that we could add at least a dbtp peak in the UI near the meters (with a reset button) to memorize clipping events?

I would have to take a look at how true peaks are measured. But I have the feeling it may not be so straightforward light on resources to the point of putting one in each plugin. Maybe it is time to consider integrating one of the x42 meters to EasyEffects. Specially now that we support multiple filter instances. People who need this information could just add the meter when needed and where needed.

**Comment by bhack** [comment]
Yes we need to find a solution as it will be useful also when an user it is going to adapt a classical parameteric eq preset. He need to manually find a new gain value if we cannot automatically recompute it.

**Comment by bhack** [comment]
> Maybe it is time to consider integrating one of the x42 meters to EasyEffects. Specially now that we support multiple filter instances. People who need this information could just add the meter when needed and where needed.

Do you mean to add it as lv2 in the pipewire filter chain?

**Comment by wwmm** [comment]
>Do you mean to add it as lv2 in the pipewire filter chain?

Not with pipewire filter chain. I think we should use our won Lv2 wrapper and wrap the x42 plugin in the same way we wrapped the other LV2 plugins available in EasyEffects. In other word the idea would be to use it like the limiter, compressor, etc.

**Comment by bhack** [comment]
What do you think about https://github.com/jaakkopasanen/AutoEq/issues/388#issuecomment-1264594335 ?

**Comment by wwmm** [comment]
>What do you think about https://github.com/jaakkopasanen/AutoEq/issues/388#issuecomment-1264594335 ?

I don't know. Once I have time I will try to use other types of normalization to see what happens.

**Comment by LethalManBoob** [comment]
I know this is an old issue but convolver auto gain always causes my convolution filters to clip on some music.

**Comment by wwmm** [comment]
> I know this is an old issue but convolver auto gain always causes my convolution filters to clip on some music.

I think that a bit of manual adjustment through the plugin input and output gains is hard to avoid. Different impulse files often have very different signal levels.

Another option is using a plugin like the `Autogain` after the convolver so the final level is below 0 dB.
